Make auto-service and auto-value incremental processors (for Gradle)

tbroyer opened the pull request:
#615

oehme commented on service/pom.xml:
```diff
@@ -40,6 +40,12 @@
   </scm>
 
   <dependencies>
+    <dependency>
```
I don't think the convenience library will work, because this build uses `proc:none`. You'll have to do the registration the manual way (which isn't that much harder).

tbroyer replied:
Ha! good catch; fixed.

tbroyer commented:
@ronshapiro Is this something you'd like to see me pursuing? (particularly wrt the API additions for autovalue extensions)

ronshapiro commented:
I'll let @eamonnmcmanus comment on AutoValue. AutoService support sounds fine, though I imagine it's often not in large builds?

oehme commented:
I agree AutoService is rare in large projects. But since it's so easy to make incremental, there's really no reason not to :)

bingranl commented:
Thanks for your PR! @tbroyer I am from Google android build team. I am wondering if there is a good way to test it which can also be used for other APs. Do you have something on your end? @oehme (I know there is a gradle tooling api)

oehme commented:
What do you have in mind beyond the tooling API? The test cases would have to be processor specific.

bingranl commented:
What is the reason that tests have to be specific? Is it because you have to write functional/unit test to make sure existing functionality works? I am hoping to find a way to tell if it is incremental or not as Gradle doc mentions  "As a user you can see which annotation processors are triggering full recompilations in the --info log"

oehme commented:
I don't think it's worth testing whether you successfully registered, that's trivial and unlikely to break. The interesting part is testing whether your processor actually honors the contract (i.e. provides originating elements, doesn't use any hidden channels like java.io etc.). These kinds of tests (if you want to add them) would be processor specific.

bingranl commented:
Then what can --info log tell us? You mean even if it shows this AP is incremental, it is very likely this AP breaks some contracts and are not actually incremental?

oehme commented:
My point is that a bad processor may do things that Gradle cannot detect. So we'll run incremental, but possibly fail. That's why it's important that the processor author reads and understands the restrictions in the documentation and makes sure their processor follows these restrictions. This can be done with code review and supported by testing where necessary.

auto-value and auto-service are clearly well-behaved, they only use the annotation processing APIs as intended. But I've seen other processors that access compiler internals or use `java.io`. Those need to change their approach before they can declare themselves as incremental.

eamonnmcmanus commented:
What's the status of this PR? Is it still Work In Progress as the title says, or could we consider taking it as it stands?

eamonnmcmanus commented on value/src/main/java/com/google/auto/value/processor/AutoValueProcessor.java:
```diff
@@ -112,6 +116,23 @@ public synchronized void init(ProcessingEnvironment processingEnv) {
+        return ImmutableSet.of(OMIT_IDENTIFIERS_OPTION, IncrementalAnnotationProcessorType.ISOLATING.getProcessorOption());
+      case AGGREGATING:
+        return ImmutableSet.of(OMIT_IDENTIFIERS_OPTION, IncrementalAnnotationProcessorType.AGGREGATING.getProcessorOption());
+      case UNKNOWN:
```
I would delete this, or delete the default case and follow the switch with `throw new AssertionError(incrementalType)`.

eamonnmcmanus commented on value/src/main/java/com/google/auto/value/extension/AutoValueExtension.java:
```diff
@@ -84,6 +84,18 @@
     Set<ExecutableElement> abstractMethods();
   }
 
+  // TODO: javadoc
```
Can you add this? Don't forget to mention that the order is important.

tbroyer commented:
> What's the status of this PR? Is it still Work In Progress as the title says, or could we consider taking it as it stands?

You may want to add tests, and/or an option (or system property) to override / disable incremental processing as a whole (like it's being done for Dagger).

I've otherwise addressed your feedback.

eamonnmcmanus commented on value/src/main/java/com/google/auto/value/extension/AutoValueExtension.java:
```diff
@@ -108,6 +108,38 @@
+   *     documentation of its incremental annotation processing</a>
+   */
+  public enum IncrementalExtensionType {
+    UNKNOWN,
```
I feel like the individual enum constants should have an understandable summary of what they mean. Here's what I've written in the copy of this PR that I'm importing into Google's private repo:

```java
    /**
     * The incrementality of this extension is unknown, or it is neither aggregating nor isolating.
     */
    UNKNOWN,

    /**
     * This extension is <i>aggregating</i>, meaning that it may generate outputs based on several
     * annotated input classes and it respects the constraints imposed on aggregating processors.
     * It is unusual for AutoValue extensions to be aggregating.
     *
     * @see <
     *     href="https://docs.gradle.org/current/userguide/java_plugin.html#aggregating_annotation_processors">Gradle
     *     definition of aggregating processors</a>
     */
    AGGREGATING,

    /**
     * This extension is <i>isolating</i>, meaning roughly that its output depends on the
     * {@code @AutoValue} class and its dependencies, but not on other {@code @AutoValue} classes
     * that might be compiled at the same time. The constraints that an isolating extension must
     * respect are the same as those that Gradle imposes on an isolating annotation processor.
     *
     * @see <
     *     href="https://docs.gradle.org/current/userguide/java_plugin.html#isolating_annotation_processors">Gradle
     *     definition of aggregating processors</a>
     */
    ISOLATING
```

Does that seem right? I'm actually finding it hard to imagine what an aggregating extension might be trying to do.

tbroyer replied:
> Does that seem right?

SGreatTM :wink:

> I'm actually finding it hard to imagine what an aggregating extension might be trying to do.

IIUC, looking up another class by name, e.g using a naming rule and/or an annotation property value (or processor option).
Probably unusual, but not impossible… :man_shrugging:
